### ISSUE TITLE
Include OpenJDK in tests to be compatible with Docker

### ISSUE THREAD

**Issue by sooheelee** [broadinstitute/picard#813]
Excerpt from SLACK conversation that concluded in GATK adding OpenJDK in code tests so as to enable using the program in a Docker image. If Picard also wants to be knowingly compatible, then perhaps consider following suite. 

kshakir
[10:56 AM] 
@shlee I was a whole version off in my description OpenJDK's woes. GATK 3 ran on JDK 6, and I think skipped JDK 7 all together, going straight to JDK 8. Meanwhile, *Open*JDK 6 was the one that was backported from an early version of OpenJDK 7, causing numerous bugs in OpenJDK 6. See "Genealogy" under http://openjdk.java.net/projects/jdk6/ if you'd like a reference link for the future. Since JDK 7, *Oracle*JDK 7+ has been a superset of *Open*JDK 7+. See also: "Q: What is the difference between the source code found in the OpenJDK repository, and the code you use to build the Oracle JDK?" in https://blogs.oracle.com/henrik/entry/java_7_questions_answers
blogs.oracle.com
Java 7 Questions & Answers (Henrik on Java)
Blogs.Oracle.Com - Henrik on Java

shlee [10:58 AM] 
Have a question in turn @kshakir. Does "re-distribute"  include individual to individual? Or is it okay for me to share my Docker image with a "collaborator", e.g. using the Collaborators feature of Dockerhub?

[10:58] 
I will ask our devs if they test in OpenJDK

kshakir
[10:59 AM] 
http://blog.takipi.com/running-java-on-docker-youre-breaking-the-law/
Takipi Blog
Running Java on Docker? You're Breaking the Law | Takipi Blog
If you’re using Java on Docker you might be breaking the law by violating Oracle’s License Agreement, without even knowing it (62kB)
March 10th, 2016 at 10:15 AM

shlee [11:00 AM] 
We do not test GATK on OpenJDK

[11:02] 
“Oracle grants you a non-exclusive, non-transferable, limited license without fees to reproduce and distribute the Software, provided that (i) you distribute the Software complete and unmodified and only bundled as part of, and for the sole purpose of running, your Programs…”

shlee [11:10 AM] 
This sounds to me: I'm allowed to make my own Docker and use Oracle Java so long as I install it directly from Oracle. It looks like we can distribute it so long as our purpose is to run our own software, e.g. GATK.

[11:11] 
Geraldine is back tomorrow and she can provide more insight

lichtens [11:12 AM] 
We do install it directly from Oracle.

[11:12] 
into the docker image

shlee [11:13 AM] 
This is what I did too

lichtens [11:13 AM] 
I don't see how this is breaking any license.  @kshakir @shlee

shlee [11:14 AM] 
It is the act of my giving my docker image to you, @lichtens with just java installed that may be problematic.

[11:14] 
Your purpose should be to run our software then I may share with you?

kshakir
[11:15 AM] 
Alice can download OracleJDK for her use. But Alice cannot distribute OracleJDK in any way that would allow Bob to use Oracle JDK without accepting the license himself.

[11:15] 
In general, there's very, very little in the OracleJDK 7+ that anyone would need outside of OpenJDK 7+.

kshakir
[11:17 AM] 
uploaded this image: JDK Flow
Add Comment

lichtens [11:19 AM] 
It's not hard for me to test.

[11:19] 
I'll do it as part of my next docker-related task/

shlee [11:20 AM] 
This clarifies. Thanks @kshakir. So all we can do is provide instructions on how to install Oracle JDK to users. If their Docker base is mac-like vs linux, then they would use a different JDK. Since GATK is untested in Open JDK, we can then only recommend a mac-like base image.

lichtens [11:21 AM] 
Hey ...  @droazen @louisb  see above from Khalid.

kshakir
[11:34 AM] 
OpenJDK has a very good variety of base images/tags here: https://hub.docker.com/r/_/openjdk/

kshakir
[11:51 AM] 
@shlee Regarding "Docker base is mac-like vs linux", the flow chart was referring to general development. Docker base images are linux (except for early support for windows based docker images, but DSDE isn't even looking at that anytime soon). So even on a Mac, if one is running Docker, they should run OpenJDK-on-Linux-via-Docker-on-Mac. If they don't want to run Docker, then it's just OracleJDK-on-Mac, as there is no OpenJDK for Mac.

shlee [11:54 AM] 
Ok. This compounds the testing we have to do. Thank you for this.

shlee [11:55 AM] 
uploaded and commented on this image: Screenshot 2017-05-10 11.51.29.png
1 Comment
Shoutout to Picard

shlee [11:57 AM] 
I'm going to @vdauwera so she can see all this when she gets back.

new messages

----- Today May 11th, 2017 -----
vdauwera [12:13 AM] 
Tl;dr: welcome home! Enjoy this license/version headache ;)

[12:15] 
If I'm not mistaken our production pipeline uses a base image that runs OpenJDK.

[12:16] 
And I've selected a similar base image to build the GATK 3 dockers.

[12:18] 
We haven't had trouble with OpenJDK in a long time, and many of our users use it, so I'm comfortable supporting it.

[12:19] 
Finally, @kshakir : we did actually run Java 7 for a few versions of GATK 3; the switch to 8 was fairly recent (3.5 or 3.6 iirc)

### Instructions

- Use a **concise** yet **descriptive** title;
- Determine whether your issue is a **bug report**, a **feature request**, or a **documentation request**;
- Choose the corresponding template block below and fill it in, replacing or deleting text in italics (surrounded by `_`) as appropriate;
- Delete the other template blocks and this header.

----

### Affected tool(s)
_Tool name(s), special parameters?_

### Affected version(s)
- [ ] Latest public release version [version?]
- [ ] Latest development/master branch as of [date of test?]

### Description 
_Describe the problem below. Provide **screenshots** , **stacktrace** , **logs** where appropriate._

#### Steps to reproduce
_Tell us how to reproduce this issue. If possible, include command lines that reproduce the problem and provide a minimal test case._

#### Expected behavior
_Tell us what should happen_

#### Actual behavior
_Tell us what happens instead_

----

### Tool(s) involved
_Tool name(s), special parameters?_

### Description
_Specify whether you want a modification of an existing behavior or addition of a new capability._
_Provide **examples**, **screenshots**, where appropriate._

----

### Tool(s) involved
_Tool name(s), parameters?_

### Description 
_Describe what needs to be added or modified._

----

**Comment by sooheelee** [comment]
Putative OpenJDK error: http://gatkforums.broadinstitute.org/gatk/discussion/9632/how-to-make-mergebamalignment-work

**Comment by ktibbett** [comment]
All our picard builds now test with both oracle and openjdk
